From: "Allen.Ye" <allen.ye@mediatek.com>
Date: Thu, 8 Jun 2023 17:32:33 +0800
Subject: [PATCH] wifi: mt76: mt7996: add check for hostapd config he_ldpc

Add check for hostapd config he_ldpc.
This capabilities is checked in mcu_beacon_check_caps in 7915.

Add check for STA LDPC cap, if STA only have BCC we should not
overwrite the phy_cap with config he_ldpc.

Signed-off-by: Allen.Ye <allen.ye@mediatek.com>
---
 mt7996/mcu.c |    8 ++++++++
 1 file changed, 8 insertions(+)

--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1799,6 +1799,7 @@ int mt7996_mcu_add_rx_ba(struct mt7996_d
 
 static void
 mt7996_mcu_sta_he_tlv(struct sk_buff *skb,
+		      struct ieee80211_bss_conf *link_conf,
 		      struct ieee80211_link_sta *link_sta,
 		      struct mt7996_vif_link *link)
 {
@@ -1820,6 +1821,11 @@ mt7996_mcu_sta_he_tlv(struct sk_buff *sk
 		he->he_phy_cap[i] = elem->phy_cap_info[i];
 	}
 
+	if (link_conf->vif->type == NL80211_IFTYPE_AP &&
+	    (elem->phy_cap_info[1] & IEEE80211_HE_PHY_CAP1_LDPC_CODING_IN_PAYLOAD))
+		u8p_replace_bits(&he->he_phy_cap[1], link_conf->he_ldpc,
+				 IEEE80211_HE_PHY_CAP1_LDPC_CODING_IN_PAYLOAD);
+
 	mcs_map = link_sta->he_cap.he_mcs_nss_supp;
 	switch (link_sta->bandwidth) {
 	case IEEE80211_STA_RX_BW_160:
@@ -2861,7 +2867,7 @@ int mt7996_mcu_add_rate_ctrl(struct mt79
 	 * update sta_rec_he here.
 	 */
 	if (changed)
-		mt7996_mcu_sta_he_tlv(skb, link_sta, link);
+		mt7996_mcu_sta_he_tlv(skb, link_conf, link_sta, link);
 
 	/* sta_rec_ra accommodates BW, NSS and only MCS range format
 	 * i.e 0-{7,8,9} for VHT.
@@ -3261,7 +3267,7 @@ int mt7996_mcu_add_sta(struct mt7996_dev
 		mt7996_mcu_sta_amsdu_tlv(dev, skb, link_conf->vif, link_sta,
 					 msta_link);
 		/* starec he */
-		mt7996_mcu_sta_he_tlv(skb, link_sta, link);
+		mt7996_mcu_sta_he_tlv(skb, link_conf, link_sta, link);
 		/* starec he 6g*/
 		mt7996_mcu_sta_he_6g_tlv(skb, link_sta);
 		/* starec eht */
